### DISCUSSION TITLE
Webpack 5 is consistently ~30% slower for all our packages compared to Webpack 4

### DISCUSSION THREAD

**Discussion by JonWallsten** [webpack/webpack#12475] [Q&A]
After migrating from Webpack 4 to Webpack 5 i started to notice increased build times.
I have not measured anything in development mode, but I have made lots of testing in production mode (less plugins and bloat). 
I have measured with different settings and plugins and Webpack 5 is consistently around at least 30% slower for both our core library build and our Angular app bild.
I have not yet figured out a way to decrease the time in the Webpack 5 build with any modifications. 
I have tried to remove all loaders besides ts-loader, which is of course essential for this to run at all.

**Before I create a bug report I wanted to start a discussion about possible sources of this issue.**

Measurements:

**Core Library**
Version | v4 (sec) | v5 (sec)
-- | -- | --
  | 19,113 | 25,722
  | 19,405 | 23,968
  | 18,218 | 24,559
  | 18,489 | 23,877
Avg | 18,912 | 24,5315
Diff (%) |   | 29,71394

**Angular App**
Version | v4 (sec) | v5 (sec)
-- | -- | --
  | 28,603 | 38,581
  | 28,773 | 37,632
  | 28,582 | 38,781
  | 28,084 | 40,196
Avg | 28,65267 | 38,7975
Diff (%) |   | 35,40624

**My config for Webpack 4**
https://pastebin.com/ZbUG3fp4

** Versions for Webpack 4**
```
webpack: 4.44.1
webpack-cl: 3.3.12
terser-webpack-plugin: 4.2.3
ts-loader: 8.0.11
typescript: 4.0.5
node: 14.15.1
os: Windows 10 v.1809
```

**Build log for Webpack 4**
```
> node --max_old_space_size=8192 ../../node_modules/webpack/bin/webpack.js "--config" "config/webpack.prod.ts"

Hash: c134fe72f1951c68154d 
Version: webpack 4.44.1
Time: 18489ms
Built at: 2021-01-22 11:53:43
                                                                    Asset        Size  Chunks                    Chunk Names
                                                                 api.d.ts       3 KiB          [emitted]
								 ....
								 ....
								 ....
Entrypoint index [big] = index.js
  [1] ./src/***log.class.ts 37 KiB {0} [built]
  [2] ./src/utils/index.ts 22.8 KiB {0} [built]
  [3] ./src/types/***pojodata.ts 32.3 KiB {0} [built]
  [4] ./src/constants.ts 10 KiB {0} [built]
  [5] ./src/pojo/***pojoapi.ts 40.9 KiB {0} [built]
  [6] ./src/utils/object.ts 5.67 KiB {0} [built]
  [7] ./src/metadata/***metadataapi.class.ts 10.4 KiB {0} [built]
  [8] ./src/utils/decorators.ts 3.47 KiB {0} [built]
 [10] ./src/types/***types.ts 4.3 KiB {0} [built]
 [11] ./src/pojo/typeguards.ts 18.4 KiB {0} [built]
 [12] ./src/transactions/transactionutils.ts 30.8 KiB {0} [built]
 [13] ./src/utils/json.ts 1.02 KiB {0} [built]
 [14] ./src/appinfo.ts 3.78 KiB {0} [built]
 [15] ./src/common.ts 16.6 KiB {0} [built]
[143] ./src/index.ts 13.8 KiB {0} [built]
    + 235 hidden modules

WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  index.js (1.07 MiB)

WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.      
Entrypoints:
  index (1.07 MiB)
      index.js


WARNING in webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/
```

**My config for Webpack 5**
https://pastebin.com/5QMP7XPJ

** Versions for Webpack 5**
```
webpack: 5.16.0
webpack-cl: 4.4.0
terser-webpack-plugin: 5.1.1
ts-loader: 8.0.14
typescript: 4.1.3
node: 14.15.1
os: Windows 10 v.1809
```

**Build log for Webpack 5**
```
> node --max_old_space_size=8192 ../../node_modules/webpack/bin/webpack.js "--config" "config/webpack.prod.ts" "--progress" "profile"

<i> [webpack.Progress]  | 35 ms setup > compilation
<i> [webpack.Progress] 49 ms setup
<i> [webpack.Progress]  | 135 ms building > import loader ../../node_modules/ts-loader/index.js
<e> [webpack.Progress]  | 17260 ms building > entries dependencies modules
<e> [webpack.Progress] 17447 ms building
<i> [webpack.Progress]  |  | 15 ms dependencies optimization > FlagDependencyUsagePlugin
<i> [webpack.Progress]  | 20 ms sealing > dependencies optimization
<i> [webpack.Progress]  | 19 ms sealing > code generation
<w> [webpack.Progress]  |  | 5000 ms asset processing > TerserPlugin
<w> [webpack.Progress]  | 5005 ms sealing > asset processing
<w> [webpack.Progress]  |  | 2091 ms after asset optimization > ts-loader
<w> [webpack.Progress]  | 2098 ms sealing > after asset optimization
<w> [webpack.Progress] 7227 ms sealing
<i> [webpack.Progress]  | 192 ms emitting > emit
<i> [webpack.Progress] 199 ms emitting
413 assets
runtime modules 668 bytes 3 modules
modules by path ./src/ 2.52 MiB 200 modules
modules by path ../../node_modules/ 56.7 KiB
  modules by path ../../node_modules/querystring/*.js 4.51 KiB 3 modules
  modules by path ../../node_modules/url/*.js 23.1 KiB
    ../../node_modules/url/url.js 22.8 KiB [built] [code generated]
    ../../node_modules/url/util.js 314 bytes [built] [code generated]
  ../../node_modules/tslib/tslib.es6.js 10.8 KiB [built] [code generated]
  ../../node_modules/process/browser.js 5.29 KiB [built] [code generated]
  ../../node_modules/https-browserify/index.js 717 bytes [built] [code generated]
  ../../node_modules/punycode/punycode.es6.js 12.3 KiB [built] [code generated]
external "axios" 42 bytes [built] [code generated]
http (ignored) 15 bytes [built] [code generated]

WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  index.js (1010 KiB)

WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.    
Entrypoints:
  index (1010 KiB)
      index.js


WARNING in webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/
webpack 5.16.0 compiled with 3 warnings in 24920 ms
```

Please let me know if you need anything else.

**Comment by alexander-akait** [comment]
Please record profiling and put it here https://github.com/webpack/webpack/issues/12102

**Reply by JonWallsten** [reply]
Sure! I'll try to do that!
I read through that ticket, but since it was about rebuilds I couldn't tell if it was related to my issue or not.

**Reply by alexander-akait** [reply]
I think your rebuild will be also slow, check it, also do you try cache? Does it help for rebuilds and initial run (hot cache)?

**Comment by chenxsan** [comment]
In the [benchmark of esbuild](https://esbuild.github.io/), webpack 5 is indeed slower than webpack 4:

![image](https://user-images.githubusercontent.com/1091472/105491450-42a64a00-5cf1-11eb-8072-aca0c7d8d619.png)

**Reply by alexander-akait** [reply]
There is old records, now we have more caches and many optimizations, I think it is not truth right now

**Reply by chenxsan** [reply]
Actually I've sent a [pr](https://github.com/evanw/esbuild/pull/681) over there a few days ago, so the results are quite up-to-date.

**Reply by alexander-akait** [reply]
Here example of the one specific project, but improvements can be in many places

**Reply by alexander-akait** [reply]
For example, set `cache.type` to `filesystem` and look at result, on seconds and other runs it will be 2-3 s :smile:

**Reply by chenxsan** [reply]
Yeah, the benchmark runs a single build from scratch, without any caches.

**Reply by JonWallsten** [reply]
@alexander-akait: I tried to filesystem cache. Second build took 24 seconds, the rest of the build was < 1 sec, but the my changes was not included in the dist. And the dist file according to the timestamp was not updated. Not sure if I do something wrong or if it's a bug.

**Reply by alexander-akait** [reply]
Something wrong, it should be very fast, doesn't you have warnings/errors? Does you create repo?

**Reply by JonWallsten** [reply]
@alexander-akait: I did some more tries today. I used cache: 'filesystem' for all libraries and apps in our monorepo.
When I changed the core package (used by all apps and all other libraries the change would not propagate correctly to the apps. 
The app complained about this:
```
Error: packages\web-app-edit\src\main.ts:6:10 - error TS2305: Module '"../../web-lib-core/dist"' has no exported member 'testExport'. 
import { testExport } from '@oas/web-lib-core';
```
The build of `web-lib-core` did not generate any errors or warnings.

When looking deeper into this I can see that the code actually seems to run, so runtime is not the issue. The Typecheck probably is.
If I look in the `./dist` folder I can see that the index.js bundle file is updated, but the index.d.ts isn't.
![image](https://user-images.githubusercontent.com/5993909/105705939-2da80180-5f11-11eb-8b4e-2240adcd2c2d.png)

To the loader (ts-loader in this case) also need to support the cache for this to work?

**Reply by alexander-akait** [reply]
You try to imporve `testExport`, but it is not exported from `web-lib-core`, I can't help because I can't look at core, all loaders cached when module was built without warnings or errors

**Reply by JonWallsten** [reply]
`testExport` is exported, I can see it in the `index.js` file. But the `index.d.ts` is not updated with the type information, so the compiler complains. Since this error can be Webpack, ts-loader or TypeScript it's hard to know where to look. Maybe it doesn't generate new definition files because it's cached somehow?

Edit: I think I will skip caching for now and focus on getting a profile for the original bug.
I turned of the cache and the issue persists. So it's probably something wrong with Webpack 5 + ts-loader.

**Reply by alexander-akait** [reply]
Maybe bug on `ts-loader` side, but I don't see the same problem(s) for other, so maybe something wrong with configuration(s)

**Reply by JonWallsten** [reply]
@alexander-akait: So I've been debugging for hours now. It seems like the `afterProcessAssets` hook is never called on the subsequent builds, so the callback that writes the definition files are never called:
https://github.com/TypeStrong/ts-loader/blob/268bc6932cc636df29b5b65c23d9033bea7c83a8/src/instances.ts#L340

I guess if it was Webpack's fault it would be a bigger issue. Do you have any idea where to start digging? I have no idea what I'm doing...

**Reply by alexander-akait** [reply]
I think better to open an issue in `ts-loader`, maybe some limitations and something misleading in docs or maybe you are right and it is just a bug. If you provide minimum reproducible test repo in the issue I will help to fix

**Reply by JonWallsten** [reply]
I have opened this one:
https://github.com/TypeStrong/ts-loader/issues/1243

**Reply by JonWallsten** [reply]
It seems like they are using `afterProcessAssets` instead of `processAssets` because they can't import `Compilation.PROCESS_ASSETS_STAGE_ADDITIONS` since they also support Webpack 4.

**Reply by alexander-akait** [reply]
But they can do it :smile:

**Reply by JonWallsten** [reply]
They I try to figure it out...
This is the comment I found.
```
// It may be better to add assets at the processAssets stage (https://webpack.js.org/api/compilation-hooks/#processassets)
// This requires Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL, which does not exist in webpack4
// Consider changing this when ts-loader is built using webpack5
```

**Reply by alexander-akait** [reply]
They need to fix it... What is the problem to do it?

**Reply by JonWallsten** [reply]
So `Compilation` does exists in Webpack 4 as well?
`const { Compilation } = require('webpack');` would work? 
Or would it just be undefined?

Edit: Never mind. Got it to work.

**Reply by JonWallsten** [reply]
![image](https://user-images.githubusercontent.com/5993909/105737110-d5cfc180-5f35-11eb-8e45-ea32f523c912.png)
The callback is never called on subsequent builds, only the first. Can this be a Webpack issue? Or did I implement it wrong?
The `afterCompile` hook a few lines prior is called.

**Reply by alexander-akait** [reply]
No problem with webpack, all plugins which uses `processAssets` works fine in watch mode

**Reply by JonWallsten** [reply]
Then I'm out of luck. What could cause a hook to only be called once, but other hooks every time?

**Reply by alexander-akait** [reply]
Nothing, hooks called on each compilation

**Reply by JonWallsten** [reply]
![image](https://user-images.githubusercontent.com/5993909/105738880-ce111c80-5f37-11eb-9d3c-a8afb9b302cd.png)
Seems like one hook is registered on the compiler and one on the compilation. So the compilation is not persistent but the compiler is?
That seems correct when checking the documentation. But can the compilation be outdated or discarded? And that's why the hook is never called again?

**Reply by alexander-akait** [reply]
Example of usage https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/src/index.js#L623 - you need use `compilation.hooks.processAssets` inside `compiler.hooks.compilation`

**Reply by JonWallsten** [reply]
Thank you!

**Reply by JonWallsten** [reply]
![image](https://user-images.githubusercontent.com/5993909/105741432-95267700-5f3a-11eb-9949-bfb706384aeb.png)
Something like this? It's never called so I guess they tap in to late in the process. I need to dig a bit more.

**Reply by alexander-akait** [reply]
Please create reproducible test repo in the issue and I will help, I hard for me helping in this way

**Reply by JonWallsten** [reply]
IT WORKS! 
It's ugly code, but I'll figure that out later!

![image](https://user-images.githubusercontent.com/5993909/105742944-67423200-5f3c-11eb-991f-4abfeb8520ed.png)

Thank you so much for you patience and help! And sorry @chenxsan for all the notifications.

**Reply by sokra** [reply]
One should never tap a hook within a loader. A loader is invoked zero to many many times and would add more and more plugins to the compiler. Or it even never adds the plugin when the module is cached. Only plugins should tap into hooks.

**Reply by JonWallsten** [reply]
@johnnyreilly: see above

**Reply by johnnyreilly** [reply]
This is interesting. `ts-loader` has always done exactly this:

https://github.com/TypeStrong/ts-loader/blob/master/src/instances.ts#L324

Where you have a loader that requires tap-ing to work.... What approach can be taken?

**Reply by johnnyreilly** [reply]
Incidentally, we only tap in the initial call to the loader - so multiple taps etc should not be an issue I think.

**Reply by alexander-akait** [reply]
@johnnyreilly also you can use global option like `hasHooks` and set it to `true` to avoid multiple taps

**Reply by sokra** [reply]
> What approach can be taken?

A plugin. E.g. mini-css-webpack-plugin uses a loader and a plugin.

**Reply by JonWallsten** [reply]
So the loader would handle the TS compilation and the plugin adding the assets basically?
The downside would be that you have to add both loader and plugin in the config I guess.
But this is the intended/correct way Webpack is designed for @sokra?

**Reply by alexander-akait** [reply]
Ideally `ts-loader` should be a plugin, because the plugin API has more abilities, right now `ts-loader` code looks good and works fine, so we can postpone it

**Reply by johnnyreilly** [reply]
Rewriting `ts-loader` to be a plugin instead of a loader would be cool. But it would be a broader scope of work than we're thinking about at this moment. (And maybe we'd need a new name! `ts-plugin` anybody? 😉)

Right now we'd like to start building `ts-loader` against webpack 5 and drop support for webpack 4.

**Reply by johnnyreilly** [reply]
Thanks for sharing `mini-css-webpack-plugin` @sokra

**Reply by Bnaya** [reply]
What are the conclusions here? esbuild still shows that webpack 5 is slower than 4
